### PR TITLE
Fix survey summarizer bug take 2

### PR DESCRIPTION
**Pull request by mehalshah**

### CODE REVIEW

**Comment by aoby** [comment]
This is a large PR. Can I please get some context / summary, or is there a way to split it up for easier review?

File: apps/src/code-studio/pd/workshop_dashboard/components/survey_results/free_response_section.jsx [hunk 1]
```diff
@@ -29,13 +24,27 @@ export default class FreeResponseSection extends React.Component {
...
+    return answers.map((answer, i) => this.renderBullet(answer, i));
+  }
+
+  renderBullet(text, key) {
+    if (_.trim(text)) {
       return (
-        <li key={i}>
-          {answer}
+        <li key={key}>
+          {text}
```
Comment (aoby) [inline apps/src/code-studio/pd/workshop_dashboard/components/survey_results/free_response_section.jsx]: Do we want to render the un-trimmed text?

File: dashboard/app/controllers/api/v1/pd/workshop_organizer_survey_report_controller.rb [hunk 2]
```diff
@@ -12,10 +12,14 @@ def index
...
-      )
-      survey_report.merge!(facilitator_scores)
+      workshops = ::Pd::Workshop.where(course: params[:course], organizer_id: current_user.id).in_state(::Pd::Workshop::STATE_ENDED)
+
+      survey_report[:all_my_workshops_for_course] = get_score_for_workshops(workshops)
+
+      workshops.flat_map(&:facilitators).uniq.each do |facilitator|
+        scores_for_facilitator = get_score_for_workshops(workshops.facilitated_by(facilitator), facilitator_name_filter: facilitator.name)
+        survey_report.merge!({facilitator.name => scores_for_facilitator})
+      end
```
Comment (aoby) [inline dashboard/app/controllers/api/v1/pd/workshop_organizer_survey_report_controller.rb]: This seems like a lot of logic in a controller. Can we move this into a model or lib class and add unit tests?

I'd like to see something more like:
```ruby
survey_report = generate_survey_report(
  course: params[:course], 
  organizer_id: current_user.id
)
``` 
or
```ruby
survey_report = SurveyReport.new(
  course: params[:course],
  organizer_id: current_user.id
).to_h
```
with all the internals concerned with downloading aggregate results from S3, and querying the correct workshops, etc, encapsulated in that function / class in an organized and tested way.

Maybe the existing `WorkshopScoreSummarizer` is the appropriate place? As it stands I'm also unclear what this module does. We have this lib mixin called summarizer, but it's not used to generate the summary? Or rather it's used to generate parts of the summary, but then other implementation-specific logic such as the S3 download is outside in the controller action?

File: dashboard/app/controllers/api/v1/pd/workshop_survey_report_controller.rb [hunk 3]
```diff
@@ -10,12 +10,12 @@ class WorkshopSurveyReportController < ReportControllerBase
...
 
-      survey_report[:this_workshop] = get_score_for_workshops([@workshop], include_free_responses: true)
+      survey_report[:this_workshop] = get_score_for_workshops([@workshop], include_free_responses: true, facilitator_name_filter: facilitator_name_filter)
       all_my_workshops = (params[:organizer_view] ? Pd::Workshop.organized_by(current_user) : Pd::Workshop.facilitated_by(current_user)).where(course: @workshop.course).in_state(Pd::Workshop::STATE_ENDED)
-      survey_report[:all_my_workshops_for_course] = get_score_for_workshops(all_my_workshops)
+      survey_report[:all_my_workshops_for_course] = get_score_for_workshops(all_my_workshops, facilitator_name_filter: facilitator_name_filter)
 
       aggregate_for_all_workshops = JSON.parse(AWS::S3.download_from_bucket('pd-workshop-surveys', "aggregate-workshop-scores-#{CDO.rack_env}"))
-      survey_report[:all_workshops_for_course] = aggregate_for_all_workshops[@workshop.course].symbolize_keys
+      survey_report[:all_workshops_for_course] = aggregate_for_all_workshops[@workshop.course].try(&:symbolize_keys) || {}
```
Comment (aoby) [inline dashboard/app/controllers/api/v1/pd/workshop_survey_report_controller.rb]: here too. This is a lot of logic in a controller. This should be in a model or lib class, split up into smaller units with tests. 

This looks like a bandaid for our known broken implementation, and from this I have little confidence we have fixed it.

File: dashboard/app/controllers/api/v1/pd/workshop_survey_report_controller.rb [hunk 4]
```diff
@@ -8,14 +8,15 @@ class WorkshopSurveyReportController < ReportControllerBase
...
-      survey_report[:all_workshops_for_course] = aggregate_for_all_workshops[@workshop.course].symbolize_keys
+      all_my_workshops = params[:organizer_view] ? Pd::Workshop.organized_by(current_user) : Pd::Workshop.facilitated_by(current_user)
+      all_my_completed_workshops = all_my_workshops.where(course: @workshop.course).in_state(Pd::Workshop::STATE_ENDED)
+
+      survey_report = generate_summary_report(
+        workshop: @workshop,
+        workshops: all_my_completed_workshops,
+        course: @workshop.course,
+        facilitator_name: facilitator_name_filter
+      )
```
Comment (aoby) [inline dashboard/app/controllers/api/v1/pd/workshop_survey_report_controller.rb:19]: Much cleaner!

File: dashboard/lib/api/v1/pd/workshop_score_summarizer.rb [hunk 5]
```diff
@@ -59,150 +59,199 @@ module Api::V1::Pd::WorkshopScoreSummarizer
...
     :things_facilitator_could_improve_s
   ].freeze
 
-  def get_score_for_workshops(workshops, facilitator_breakdown: false, include_free_responses: false)
-    report_rows = Hash.new(0).merge({number_teachers: 0, response_count: 0})
+  FACILITATOR_SPECIFIC_MULTIPLE_CHOICE_QUESTIONS = FACILITATOR_SPECIFIC_QUESTIONS - FREE_RESPONSE_QUESTIONS
+  FACILITATOR_SPECIFIC_FREE_RESPONSE_QUESTIONS = FACILITATOR_SPECIFIC_QUESTIONS & FREE_RESPONSE_QUESTIONS
 
-    response_count = 0
+  def generate_summary_report(workshop: nil, workshops:, course:, facilitator_name: nil, facilitator_breakdown: false)
```
Comment (aoby) [inline dashboard/lib/api/v1/pd/workshop_score_summarizer.rb]: add YARD doc summary w `@param` descriptions

File: dashboard/lib/api/v1/pd/workshop_score_summarizer.rb [hunk 6]
```diff
@@ -59,150 +59,199 @@ module Api::V1::Pd::WorkshopScoreSummarizer
...
-        free_responses[question] = []
-      end
-    end
+    survey_report
+  end
 
-    workshops.each do |workshop|
-      enrollment_ids = workshop.enrollments.pluck(:id)
+  def get_score_for_workshops(workshops:, include_free_responses:, facilitator_name_filter:)
+    facilitators = workshops.flat_map(&:facilitators).map(&:name)
```
Comment (aoby) [inline dashboard/lib/api/v1/pd/workshop_score_summarizer.rb]: does this need a `.uniq`? What if we have multiple workshops for a given facilitator?

File: dashboard/lib/api/v1/pd/workshop_score_summarizer.rb (hunk 5 above)
Comment (mehalshah) [reply dashboard/lib/api/v1/pd/workshop_score_summarizer.rb]: Will do

File: dashboard/lib/api/v1/pd/workshop_score_summarizer.rb (hunk 6 above)
Comment (mehalshah) [reply dashboard/lib/api/v1/pd/workshop_score_summarizer.rb]: Will add

**Pull request merged**